<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="uwot">
<title>UMAP Examples • uwot</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="UMAP Examples">
<meta property="og:description" content="uwot">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">uwot</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.2.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/uwot.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/index.html">Articles</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jlmelville/uwot/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>UMAP Examples</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jlmelville/uwot/blob/HEAD/vignettes/articles/umap-examples.Rmd" class="external-link"><code>vignettes/articles/umap-examples.Rmd</code></a></small>
      <div class="d-none name"><code>umap-examples.Rmd</code></div>
    </div>

    
    
<p><em>December 29 2018</em> New, better settings for t-SNE, better
plots and a couple of new datasets. Removed neighborhood preservation
values until I’ve double checked they are working correctly.</p>
<p>Here are some examples of the output of <code>uwot</code>’s
implementation of UMAP, compared to t-SNE output. As you will see,
UMAP’s output results in more compact, separated clusters compared to
t-SNE.</p>
<div class="section level2">
<h2 id="data-preparation">Data preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h2>
<p>For details on the datasets, follow their links. Somewhat more detail
is also given in the <a href="https://jlmelville.github.io/smallvis/datasets.htm" class="external-link">smallvis
documentation</a>. <code>iris</code> you already have if you are using
R. <code>s1k</code> is part of the <a href="https://github.com/jlmelville/sneer" class="external-link">sneer</a> package.
<code>frey</code>, <code>oli</code>, <code>mnist</code>,
<code>fashion</code>, <code>kuzushiji</code>, <code>norb</code> and
<code>cifar10</code> can be downloaded via <a href="https://github.com/jlmelville/snedata" class="external-link">snedata</a>.
<code>coil20</code> and <code>coil100</code> can be fetched via <a href="https://github.com/jlmelville/coil20" class="external-link">coil20</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mnist</span> <span class="op">&lt;-</span> <span class="fu">snedata</span><span class="fu">::</span><span class="fu">download_mnist</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For some functions we need to strip out non-numeric columns and convert data to matrix</span></span>
<span><span class="va">x2m</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">methods</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/methods/is.html" class="external-link">is</a></span><span class="op">(</span><span class="va">X</span>, <span class="st">"matrix"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">X</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">is.numeric</span>, <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">logical</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">m</span> <span class="op">&lt;-</span> <span class="va">X</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">m</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>At the time I generated this document (late December 2018), the
<code>kuzushiji</code> dataset had some duplicate and all-black images
that needed filtering. This seems to have been remedied as of early
February 2019. I re-ran both UMAP and t-SNE on the fixed dataset, but
the results weren’t noticeably different. For the record, the clean-up
routines I ran were:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove all-black images in Kuzushiji MNIST (https://github.com/rois-codh/kmnist/issues/1)</span></span>
<span><span class="va">kuzushiji</span> <span class="op">&lt;-</span> <span class="va">kuzushiji</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu">x2m</span><span class="op">(</span><span class="va">kuzushiji</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">sum</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># Remove duplicate images (https://github.com/rois-codh/kmnist/issues/5)</span></span>
<span><span class="va">kuzushiji</span> <span class="op">&lt;-</span> <span class="va">kuzushiji</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="fu">x2m</span><span class="op">(</span><span class="va">kuzushiji</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="umap-settings">UMAP settings<a class="anchor" aria-label="anchor" href="#umap-settings"></a>
</h2>
<p>For UMAP, I stick with the defaults, with the exception of
<code>iris</code>, <code>coil20</code>, and <code>coil100</code> and
<code>norb</code>. The spectral initialization with the default
<code>n_neighbors</code> leads to disconnected components, which can
lead to a poor global picture of the data. The Python UMAP
implementation goes to fairly involved lengths to ameliorate theses
issues, but <code>uwot</code> does not.</p>
<p>For these datasets, a perfectly good alternative that provides a
global initialization is to use the first two components from PCA,
scaled so their standard deviations are initially 1e-4 (via
<code>init = "spca"</code>). This usually results in an embedding which
isn’t too different from starting via the raw PCA but is more compact,
i.e. less space between clusters. For visualizations, as long as the
relative orientation and rough distances between clusters are
maintained, the exact distances between them are not that interesting to
me.</p>
<p>Dimensionality was reduced to 100 by applying PCA to the data. It’s
commonly applied to data before or as part of t-SNE, so to avoid any
differences from this pre-processing, I applied PCA to the UMAP data. I
used 100 components, rather than the usual 50 just to be on the safe
side. This seemed to have no major effect on the resulting
visualizations.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For iris </span></span>
<span><span class="va">iris_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/umap.html">umap</a></span><span class="op">(</span><span class="va">iris</span>, init <span class="op">=</span> <span class="st">"spca"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Small datasets (s1k, oli, frey)</span></span>
<span><span class="va">s1k_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/umap.html">umap</a></span><span class="op">(</span><span class="va">s1k</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Big datasets (mnist, fashion, kuzushiji)</span></span>
<span><span class="va">mnist_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/umap.html">umap</a></span><span class="op">(</span><span class="va">mnist</span>, pca <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># norb, coil20 and coil100, tasic2018</span></span>
<span><span class="va">coil20_umap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/umap.html">umap</a></span><span class="op">(</span><span class="va">coil20</span>, pca <span class="op">=</span> <span class="fl">100</span>, init <span class="op">=</span> <span class="st">"spca"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="t-sne-settings">t-SNE settings<a class="anchor" aria-label="anchor" href="#t-sne-settings"></a>
</h2>
<p>The <a href="https://cran.r-project.org/package=Rtsne" class="external-link">Rtsne</a>
package was used for the t-SNE calculations, except for the
<code>iris</code> dataset, proving troublesome once again. This time
it’s because <code>Rtsne</code> doesn’t allow for duplicates. For
<code>iris</code> only, I used the <a href="https://github.com/jlmelville/smallvis" class="external-link">smallvis</a> package.</p>
<p>For t-SNE, I also employ the following non-defaults:</p>
<ul>
<li>
<code>perplexity = 15</code>, which is closer to the neighborhood
size used by UMAP</li>
<li>as mentioned in the UMAP settings section, for large datasets,
<code>initial_dims = 100</code>, which applies PCA to the input, keeping
the first 100 components, rather than the usual <code>50</code> to
minmize distortion from the initial PCA.</li>
<li>As <code>uwot</code> also makes use of <code>irlba</code>, there is
no reason not to use <code>partial_pca</code>.</li>
</ul>
<p>It would be nice to use the same initial coordinates for both
methods, but unfortunately <code>Rtsne</code> doesn’t apply early
exaggeration with user-supplied input. Without early exaggeration, t-SNE
results aren’t as good, especially with larger datasets. Therefore the
t-SNE plots use a random initialization.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For iris only</span></span>
<span><span class="va">iris_tsne</span> <span class="op">&lt;-</span> <span class="fu">smallvis</span><span class="fu">::</span><span class="fu">smallvis</span><span class="op">(</span><span class="va">iris</span>, perplexity <span class="op">=</span> <span class="fl">15</span>, Y_init <span class="op">=</span> <span class="st">"rand"</span>, exaggeration_factor <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Small datasets (s1k, oli, frey)</span></span>
<span><span class="va">s1k_tsne</span> <span class="op">&lt;-</span> <span class="fu">Rtsne</span><span class="fu">::</span><span class="fu">Rtsne</span><span class="op">(</span><span class="fu">x2m</span><span class="op">(</span><span class="va">s1k</span><span class="op">)</span>, perplexity <span class="op">=</span> <span class="fl">15</span>, initial_dims <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                       partial_pca <span class="op">=</span> <span class="cn">TRUE</span>, exaggeration_factor <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Big datasets (coil20, coil100, mnist, fashion, kuzushiji etc.)</span></span>
<span><span class="va">mnist_tsne</span> <span class="op">&lt;-</span> <span class="fu">Rtsne</span><span class="fu">::</span><span class="fu">Rtsne</span><span class="op">(</span><span class="fu">x2m</span><span class="op">(</span><span class="va">mnist</span><span class="op">)</span>, perplexity <span class="op">=</span> <span class="fl">15</span>, initial_dims <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                       partial_pca <span class="op">=</span> <span class="cn">TRUE</span>, exaggeration_factor <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h3>
<p>For visualization, I used the <a href="https://github.com/jlmelville/vizier" class="external-link">vizier</a> package. The
plots are colored by class membership (there’s an obvious choice for
every dataset considered), except for <code>frey</code>, where the
points are colored according to their position in the sequence of
images.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">embed_img</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">X</span>, <span class="va">Y</span>, <span class="va">k</span> <span class="op">=</span> <span class="fl">15</span>, <span class="va">...</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">args</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span>  <span class="va">args</span><span class="op">$</span><span class="va">coords</span> <span class="op">&lt;-</span> <span class="va">Y</span></span>
<span>  <span class="va">args</span><span class="op">$</span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">X</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="fu">vizier</span><span class="fu">::</span><span class="va">embed_plot</span>, <span class="va">args</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">embed_img</span><span class="op">(</span><span class="va">iris</span>, <span class="va">iris_umap</span>, pc_axes <span class="op">=</span> <span class="cn">TRUE</span>, equal_axes <span class="op">=</span> <span class="cn">TRUE</span>, alpha_scale <span class="op">=</span> <span class="fl">0.5</span>, title <span class="op">=</span> <span class="st">"iris UMAP"</span>, cex <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>For UMAP, where non-default initialization was used, it’s noted in
the title of the plot (e.g. “(spca)”).</p>
</div>
</div>
<div class="section level2">
<h2 id="iris">iris<a class="anchor" aria-label="anchor" href="#iris"></a>
</h2>
<p>The standard <code>iris</code> dataset, known and loved by all.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/iris_umap.png" alt="iris UMAP (spca)"></td>
<td align="center"><img src="img/examples/iris_tsne.png" alt="iris t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="s1k">s1k<a class="anchor" aria-label="anchor" href="#s1k"></a>
</h2>
<p>A 9-dimensional fuzzy simplex, which I created for testing t-SNE and
related methods, original in the <a href="https://github.com/jlmelville/sneer" class="external-link">sneer</a> package.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/s1k_umap.png" alt="s1k UMAP"></td>
<td align="center"><img src="img/examples/s1k_tsne.png" alt="s1k t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="oli">oli<a class="anchor" aria-label="anchor" href="#oli"></a>
</h2>
<p>The <a href="https://cam-orl.co.uk/facedatabase.html" class="external-link">ORL database of
faces</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/oli_umap.png" alt="oli UMAP"></td>
<td align="center"><img src="img/examples/oli_tsne.png" alt="oli t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="frey">frey<a class="anchor" aria-label="anchor" href="#frey"></a>
</h2>
<p>Images of Brendan Frey’s face, as far as I know originating from a
page belonging to <a href="https://cs.nyu.edu/~roweis/data.html" class="external-link">Saul
Roweis</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/frey_umap.png" alt="frey UMAP"></td>
<td align="center"><img src="img/examples/frey_tsne.png" alt="frey t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="isofaces">isofaces<a class="anchor" aria-label="anchor" href="#isofaces"></a>
</h2>
<p>Yet more faces, this time the dataset used in <a href="http://web.mit.edu/cocosci/isomap/datasets.html" class="external-link">Isomap</a>,
consisting of images of the same face under different rotations and
lighting conditions. Unfortunately, it’s no longer available at the MIT
website, but it can be found via <a href="https://web.archive.org/web/20160913051505/http://isomap.stanford.edu/face_data.mat.Z" class="external-link">the
Wayback Machine</a>. I wrote <a href="https://gist.github.com/jlmelville/339dfeb80c3e836e887d70a37679b244" class="external-link">a
gist for processing the data in R</a>.</p>
<p>In the images below the points are colored by the first pose
angle.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/isofaces_umap.png" alt="isofaces UMAP"></td>
<td align="center"><img src="img/examples/isofaces_tsne.png" alt="isofaces t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="coil20">coil20<a class="anchor" aria-label="anchor" href="#coil20"></a>
</h2>
<p>The <a href="http://www.cs.columbia.edu/CAVE/software/softlib/coil-20.php" class="external-link">COIL-20
Columbia Object Image Library</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/coil20_umap.png" alt="coil20 UMAP (spca)"></td>
<td align="center"><img src="img/examples/coil20_tsne.png" alt="coil20 t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="coil100">coil100<a class="anchor" aria-label="anchor" href="#coil100"></a>
</h2>
<p>The <a href="http://www.cs.columbia.edu/CAVE/software/softlib/coil-100.php" class="external-link">COIL-100
Columbia Object Image Library</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/coil100_umap.png" alt="coil100 UMAP (spca)"></td>
<td align="center"><img src="img/examples/coil100_tsne.png" alt="coil100 t-SNE"></td>
</tr></tbody>
</table>
<p>The UMAP results are rather hard to visualize on a static plot. If
you could pan and zoom around, you would see that the rather indistinct
blobs are mainly correctly preserved loops. This is an example where
choosing a different value of the <code>a</code> and <code>b</code>
parameters would be a good idea. The t-UMAP variant, available as
<code>tumap</code> uses <code>a = 1</code>, <code>b = 1</code>
(effectively using the same Cauchy kernel as t-SNE) does a better job
here and is shown below on the left. The correct choice of parameters is
also important for t-SNE. On the right is the t-SNE result with a
default <code>perplexity = 50</code>, which does not retain as much of
the loop structure as the lower perplexity:</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/coil100_tumap.png" alt="coil100 t-UMAP (spca)"></td>
<td align="center"><img src="img/examples/coil100_tsne_perp50.png" alt="coil100 t-SNE (perplexity 50)"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="swiss-roll">swiss roll<a class="anchor" aria-label="anchor" href="#swiss-roll"></a>
</h2>
<p>The <a href="http://web.mit.edu/cocosci/isomap/datasets.html" class="external-link">Swiss
Roll</a> data used in Isomap. A famous dataset, but perhaps not that
representative of typical real world datasets. t-SNE is know to not
handle this well, but UMAP makes an impressive go at unfolding it.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/swiss_umap.png" alt="swiss UMAP"></td>
<td align="center"><img src="img/examples/swiss_tsne.png" alt="swiss t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="mnist">mnist<a class="anchor" aria-label="anchor" href="#mnist"></a>
</h2>
<p>The <a href="http://yann.lecun.com/exdb/mnist/" class="external-link">MNIST database of
handwritten digits</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/mnist_umap.png" alt="mnist UMAP"></td>
<td align="center"><img src="img/examples/mnist_tsne.png" alt="mnist t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="fashion">fashion<a class="anchor" aria-label="anchor" href="#fashion"></a>
</h2>
<p>The <a href="https://github.com/zalandoresearch/fashion-mnist" class="external-link">Fashion MNIST
database</a>, images of fashion objects.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/fashion_umap.png" alt="fashion UMAP"></td>
<td align="center"><img src="img/examples/fashion_tsne.png" alt="fashion t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="kuzushiji-kmnist">kuzushiji (KMNIST)<a class="anchor" aria-label="anchor" href="#kuzushiji-kmnist"></a>
</h2>
<p>The <a href="https://github.com/rois-codh/kmnist" class="external-link">Kuzushiji MNIST
database</a>, images of cursive Japanese handwriting.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/kmnist_umap.png" alt="kuzushiji UMAP"></td>
<td align="center"><img src="img/examples/kmnist_tsne.png" alt="kuzushiji t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="norb">norb<a class="anchor" aria-label="anchor" href="#norb"></a>
</h2>
<p>The <a href="https://cs.nyu.edu/~ylclab/data/norb-v1.0-small/" class="external-link">small
NORB dataset</a>, pairs of images of 50 toys photographed at different
angles and under different lighting conditions.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/norb_umap.png" alt="norb UMAP"></td>
<td align="center"><img src="img/examples/norb_tsne.png" alt="norb t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="cifar10">cifar10<a class="anchor" aria-label="anchor" href="#cifar10"></a>
</h2>
<p>The <a href="https://www.cs.toronto.edu/~kriz/cifar.html" class="external-link">CIFAR-10
dataset</a>, consisting of 60000 32 x 32 color images evenly divided
across 10 classes (e.g. airplane, cat, truck, bird). t-SNE was applied
to CIFAR-10 in the <a href="http://jmlr.org/papers/v15/vandermaaten14a.html" class="external-link">Barnes-Hut t-SNE
paper</a>, but in the main paper, only the results after passing through
a convolutional neural network were published. t-SNE on the original
pixel data was only given in the <a href="https://lvdmaaten.github.io/publications/misc/Supplement_JMLR_2014.pdf" class="external-link">supplementary
information (PDF)</a> which is oddly hard to find a link to via JMLR or
the article itself, and in the less widely-cited <a href="https://arxiv.org/abs/1301.3342" class="external-link">preliminary investigation into BH
t-SNE</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/cifar10_umap.png" alt="cifar10 UMAP"></td>
<td align="center"><img src="img/examples/cifar10_tsne.png" alt="cifar10 t-SNE"></td>
</tr></tbody>
</table>
<p>There is an outlying orange cluster (which isn’t easy to see) in the
top right of the UMAP plot. I see the same thing in the Python
implementation, so I don’t think this is a bug in <code>uwot</code>
(although I also said that in a previous version of this page, and it
turned out there <em>was</em> a bug in <code>uwot</code>. The current
result really is closer to the Python version now, though). That cluster
of images is of some automobiles but they all seem to be variations of
the same image. The same cluster is present in the t-SNE plot (bottom
left), but is more comfortably close to the rest of the data.</p>
<p>The existence of these near-duplicates in CIFAR-10 doesn’t seem to
have been widely known or appreciated until quite recently, see for
instance <a href="https://twitter.com/colinraffel/status/1030532862930382848" class="external-link">this
twitter thread</a> and <a href="https://arxiv.org/abs/1806.00451" class="external-link">this
paper by Recht and co-workers</a>. Such manipulations are in line with
the practice of data augmentation that is popular in deep learning, but
you would need to be aware of it to avoid the test set results being
contaminated. These images seem like a good argument for applying UMAP
or t-SNE to your dataset as a way to spot this sort of thing.</p>
<p>We can get better results with UMAP by using the scaled PCA
initialization (below on the left) and by using the t-UMAP settings
(below, right):</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/cifar10_umaps.png" alt="cifar10 UMAP (spca)"></td>
<td align="center"><img src="img/examples/cifar10_tumaps.png" alt="cifar10 t-UMAP (spca)"></td>
</tr></tbody>
</table>
<p>That rogue cluster is still present (off to the lower right now), but
either image is more comparable to the t-SNE result.</p>
<p>As the visualization of CIFAR-10 isn’t very successful with UMAP or
t-SNE, here are some results using the activations of a convnet, similar
to that used in the BH t-SNE paper. For the convnet, I used a <a href="https://keras.io/" class="external-link">keras</a> implementation, taken from the <a href="https://appliedmachinelearning.blog/2018/03/24/achieving-90-accuracy-in-object-recognition-task-on-cifar-10-dataset-with-keras-convolutional-neural-networks/" class="external-link">Machine
Learning in Action blog</a>. Features were Z-scaled as carried out in
the blog, but I used 100 epochs and a batch size of 128 and I also used
the <a href="https://www.cs.toronto.edu/~tijmen/csc321/slides/lecture_slides_lec6.pdf" class="external-link">RMSprop
(PDF)</a> optimizer favored in the <a href="https://www.manning.com/books/deep-learning-with-python" class="external-link">Deep
Learning with Python book</a> (with <code>lr=1e-4</code> and
<code>decay=1e-6</code>). Without any data augmentation, this gave a
test set error of <code>0.1655</code>, slightly lower than the test set
result given in the BH t-SNE paper (which used a different architecture
and without the benefit of an extra 4 years of deep learning research).
After retraining with all 60000 images, the flattened output of the
final max-pool layer was used, giving 2048 features (the BH t-SNE paper
network had 1024 output activations). UMAP and t-SNE results are below.
For the UMAP results, I used the t-UMAP settings with scaled PCA
initialization.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/cifar10act_umap.png" alt="cifar10 activations t-UMAP"></td>
<td align="center"><img src="img/examples/cifar10act_tsne.png" alt="cifar10 activations t-SNE"></td>
</tr></tbody>
</table>
<p>Results don’t look quite as good as those in the BH t-SNE paper, but
they are still an improvement. The orange cluster of automobiles remains
as an outlier, even in the activation space. You can also see it in the
BH t-SNE paper in the lower image in Figure 5 (orange cluster at the
bottom, slightly left of center).</p>
</div>
<div class="section level2">
<h2 id="tasic2018">tasic2018<a class="anchor" aria-label="anchor" href="#tasic2018"></a>
</h2>
<p>The <code>tasic2018</code> dataset is a transcriptomics dataset of
mouse brain cell RNA-seq data from the <a href="http://celltypes.brain-map.org/rnaseq/mouse" class="external-link">Allen Brain Atlas</a>
(originally reported by <a href="http://dx.doi.org/10.1038/s41586-018-0654-5" class="external-link">Tasic and
co-workers</a>. There is gene expression data for 14,249 cells from the
primary visual cortex, and 9,573 cells from the anterior lateral motor
cortex to give a dataset of size n = 23,822 overall. Expression data for
45,768 genes were obtained in the original data, but the dataset used
here follows the pre-processing treatment of <a href="https://doi.org/10.1101/453449" class="external-link">Kobak and Berens</a> which applied
a normalization and log transformation and then only kept the top 3000
most variable genes.</p>
<p>The data can be generated from the Allen Brain Atlas website and
processed in Python by following the instructions in this <a href="https://github.com/berenslab/rna-seq-tsne/blob/master/demo.ipynb" class="external-link">Berens
lab notebook</a>. I output the data to CSV format for reading into R and
assembling into a data frame with the following extra exporting
code:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-log3k.csv"</span>, logCPM, delimiter<span class="op">=</span><span class="st">","</span>)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-areas.csv"</span>, tasic2018.areas, delimiter<span class="op">=</span><span class="st">","</span>, fmt <span class="op">=</span> <span class="st">"</span><span class="sc">%d</span><span class="st">"</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-genes.csv"</span>, tasic2018.genes[selectedGenes], delimiter<span class="op">=</span><span class="st">","</span>, fmt<span class="op">=</span><span class="st">'</span><span class="sc">%s</span><span class="st">'</span>)</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-clusters.csv"</span>, tasic2018.clusters, delimiter<span class="op">=</span><span class="st">","</span>, fmt<span class="op">=</span><span class="st">'</span><span class="sc">%d</span><span class="st">'</span>)</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-cluster-names.csv"</span>, tasic2018.clusterNames, delimiter<span class="op">=</span><span class="st">","</span>, fmt<span class="op">=</span><span class="st">'</span><span class="sc">%s</span><span class="st">'</span>)</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-cluster-colors.csv"</span>, tasic2018.clusterColors, delimiter<span class="op">=</span><span class="st">","</span>, fmt<span class="op">=</span><span class="st">'</span><span class="sc">%s</span><span class="st">'</span>)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/tasic2018_umap.png" alt="tasic2018 UMAP"></td>
<td align="center"><img src="img/examples/tasic2018_tsne.png" alt="tasic2018 t-SNE"></td>
</tr></tbody>
</table>
<p>Again, the default UMAP settings produce clusters that are bit too
well-separated to clearly see in these images. So here are the results
from using t-UMAP (the same as setting <code>a = 1, b = 1</code>), on
the left, and then using <code>a = 2, b = 2</code> on the right:</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/tasic2018_tumap.png" alt="tasic2018 t-UMAP"></td>
<td align="center"><img src="img/examples/tasic2018_umap2.png" alt="tasic2018 UMAP (a = 2, b = 2)"></td>
</tr></tbody>
</table>
</div>
<div class="section level2">
<h2 id="macosko2015">macosko2015<a class="anchor" aria-label="anchor" href="#macosko2015"></a>
</h2>
<p>Another <a href="https://doi.org/10.1016/j.cell.2015.05.002" class="external-link">transcriptomics data
set</a>, used as an example in <a href="https://github.com/pavlin-policar/openTSNE" class="external-link">openTSNE</a>. This
contains data for 44,808 cells from the mouse retina.</p>
<p>The raw data was fetched similarly to this <a href="https://github.com/hemberg-lab/scRNA.seq.datasets/blob/master/bash/macosko.sh" class="external-link">shell
script from the Hemberg Lab</a> and then the data was prepared using the
<a href="https://github.com/pavlin-policar/openTSNE/blob/master/examples/prepare_macosko_2015.ipynb" class="external-link">openTSNE
notebook by Pavlin Policar</a>. Similarly to the <code>tasic2018</code>
dataset, data was log normalized and the 3,000 most variable genes were
retained.</p>
<p>I exported the data (without the Z-scaling and PCA dimensionality
reduction), as a CSV file, e.g.:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>np.savetxt(<span class="st">"/path/to/macosko2015/macosko2015-log3k.csv"</span>, x, delimiter<span class="op">=</span><span class="st">","</span>)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="co"># Use these as column names</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>np.savetxt(<span class="st">"/path/to/macosko2015/macosko2015-genenames.csv"</span>, data.T.columns.values[gene_mask].astype(<span class="bu">str</span>), delimiter<span class="op">=</span><span class="st">","</span>, fmt <span class="op">=</span> <span class="st">"</span><span class="sc">%s</span><span class="st">"</span>)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>np.savetxt(<span class="st">"/path/to/macosko2015/macosko2015-clusterids.csv"</span>, cluster_ids.values.astype(<span class="bu">int</span>), delimiter<span class="op">=</span><span class="st">","</span>, fmt <span class="op">=</span> <span class="st">"</span><span class="sc">%d</span><span class="st">"</span>)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/macosko2015_umap.png" alt="macosko2015 UMAP"></td>
<td align="center"><img src="img/examples/macosko2015_tsne.png" alt="macosko2015 t-SNE"></td>
</tr></tbody>
</table>
<p>This is another result where the UMAP defaults might need a bit of
fiddling with if you don’t like how separated the clusters are.</p>
<p>The openTSNE results use Z-scaling of the inputs before applying
t-SNE, so below are the results for UMAP and t-SNE with Z-scaling
applied (for <code>umap</code>, pass <code>scale = "Z"</code> as an
argument so you don’t have to do it manually):</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img/examples/macosko2015_tumapz.png" alt="macosko2015 t-UMAP, Z-scaled"></td>
<td align="center"><img src="img/examples/macosko2015_tsnez.png" alt="macosko2015 t-SNE Z-scaled"></td>
</tr></tbody>
</table>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Melville.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
