<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>UMAP Examples • uwot</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><script src="../extra.js"></script><meta property="og:title" content="UMAP Examples">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">uwot</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.2.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/uwot.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jlmelville/uwot/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>UMAP Examples</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jlmelville/uwot/blob/master/vignettes/articles/umap-examples.Rmd" class="external-link"><code>vignettes/articles/umap-examples.Rmd</code></a></small>
      <div class="d-none name"><code>umap-examples.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><em>December 26 2024</em>: Images have been (mainly) reworked so that
they make better use of the space in the tables and are aligned more
consistently allowing an easier comparison of how UMAP and t-SNE differ.
The t-SNE output is also improved. I also use more default settings.
Plus more datasets.</p>
<p><em>December 29 2018</em> New, better settings for t-SNE, better
plots and a couple of new datasets. Removed neighborhood preservation
values until I’ve double checked they are working correctly.</p>
<p>Here are some examples of the output of <code>uwot</code>’s
implementation of UMAP, compared to t-SNE output. As you will see,
UMAP’s output results in more compact, separated clusters compared to
t-SNE.</p>
<div class="section level3">
<h3 id="data-preparation">Data preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h3>
<p>For details on the datasets, follow their links. Somewhat more detail
is also given in the <a href="https://jlmelville.github.io/smallvis/datasets.htm" class="external-link">smallvis
documentation</a>. <code>iris</code> you already have if you are using
R. <code>s1k</code> is part of the <a href="https://github.com/jlmelville/sneer" class="external-link">sneer</a> package.
<code>frey</code>, <code>oli</code>, <code>mnist</code>,
<code>fashion</code>, <code>kuzushiji</code>, <code>norb</code> and
<code>cifar10</code> can be downloaded via <a href="https://github.com/jlmelville/snedata" class="external-link">snedata</a>.
<code>coil20</code> and <code>coil100</code> can be fetched via <a href="https://github.com/jlmelville/coil20" class="external-link">coil20</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mnist</span> <span class="op">&lt;-</span> <span class="fu">snedata</span><span class="fu">::</span><span class="fu">download_mnist</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># For some functions we need to strip out non-numeric columns and convert data to matrix</span></span>
<span><span class="va">x2m</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">X</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu">methods</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/methods/is.html" class="external-link">is</a></span><span class="op">(</span><span class="va">X</span>, <span class="st">"matrix"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">X</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">X</span>, <span class="va">is.numeric</span>, <span class="fu"><a href="https://rdrr.io/r/base/logical.html" class="external-link">logical</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="va">m</span> <span class="op">&lt;-</span> <span class="va">X</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">m</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>At the time I generated this document (late December 2018), the
<code>kuzushiji</code> dataset had some duplicate and all-black images
that needed filtering. This seems to have been remedied as of early
February 2019. I re-ran both UMAP and t-SNE on the fixed dataset, but
the results weren’t noticeably different. For the record, the clean-up
routines I ran were:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Remove all-black images in Kuzushiji MNIST (https://github.com/rois-codh/kmnist/issues/1)</span></span>
<span><span class="va">kuzushiji</span> <span class="op">&lt;-</span> <span class="va">kuzushiji</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu">x2m</span><span class="op">(</span><span class="va">kuzushiji</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">sum</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co"># Remove duplicate images (https://github.com/rois-codh/kmnist/issues/5)</span></span>
<span><span class="va">kuzushiji</span> <span class="op">&lt;-</span> <span class="va">kuzushiji</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="fu">x2m</span><span class="op">(</span><span class="va">kuzushiji</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="settings">Settings<a class="anchor" aria-label="anchor" href="#settings"></a>
</h3>
<p>For UMAP I used default settings with the <code>umap2</code> function
with one exception: I have the pre-calculated exact nearest neighbors
for these datasets to hand, so I used those rather than approximate
nearest neighbors. For more on <code>umap2</code> see the <a href="https://jlmelville.github.io/uwot/articles/umap2.html">umap2
article</a>.</p>
<p>For t-SNE, I used the Python package <a href="https://github.com/pavlin-policar/openTSNE" class="external-link">openTSNE</a> (version
1.0.2) with its default settings.</p>
<div class="section level4">
<h4 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h4>
<p>For visualization, I used the <a href="https://github.com/jlmelville/vizier" class="external-link">vizier</a> package.</p>
</div>
</div>
<div class="section level3">
<h3 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h3>
</div>
<div class="section level3">
<h3 id="iris">iris<a class="anchor" aria-label="anchor" href="#iris"></a>
</h3>
<p>The standard <code>iris</code> dataset, known and loved by all.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Firis-umap.png" alt="iris UMAP (spca)"></td>
<td align="center"><img src="img%2Fexamples%2Firis-tsne.png" alt="iris t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="s1k">s1k<a class="anchor" aria-label="anchor" href="#s1k"></a>
</h3>
<p>A 9-dimensional fuzzy simplex, which I created for testing t-SNE and
related methods, original in the <a href="https://github.com/jlmelville/sneer" class="external-link">sneer</a> package.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fs1k-umap.png" alt="s1k UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fs1k-tsne.png" alt="s1k t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="oli-or-olivetti">oli (or olivetti)<a class="anchor" aria-label="anchor" href="#oli-or-olivetti"></a>
</h3>
<p>The <a href="https://cam-orl.co.uk/facedatabase.html" class="external-link">ORL database of
faces</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Folivetti-umap.png" alt="oli UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Folivetti-tsne.png" alt="oli t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="frey">frey<a class="anchor" aria-label="anchor" href="#frey"></a>
</h3>
<p>Images of Brendan Frey’s face, as far as I know originating from a
page belonging to <a href="https://cs.nyu.edu/~roweis/data.html" class="external-link">Saul
Roweis</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Ffrey-umap.png" alt="frey UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Ffrey-tsne.png" alt="frey t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="isofaces">isofaces<a class="anchor" aria-label="anchor" href="#isofaces"></a>
</h3>
<p>Yet more faces, this time the dataset used in <a href="http://web.mit.edu/cocosci/isomap/datasets.html" class="external-link">Isomap</a>,
consisting of images of the same face under different rotations and
lighting conditions. Unfortunately, it’s no longer available at the MIT
website, but it can be found via <a href="https://web.archive.org/web/20160913051505/http://isomap.stanford.edu/face_data.mat.Z" class="external-link">the
Wayback Machine</a>. I wrote a <a href="https://gist.github.com/jlmelville/339dfeb80c3e836e887d70a37679b244" class="external-link">gist
for processing the data in R</a>.</p>
<p>In the images below the points are colored by the left-to-right pose
angle.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fisofaces-umap.png" alt="isofaces UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fisofaces-tsne.png" alt="isofaces t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="coil20">coil20<a class="anchor" aria-label="anchor" href="#coil20"></a>
</h3>
<p>The <a href="http://www.cs.columbia.edu/CAVE/software/softlib/coil-20.php" class="external-link">COIL-20
Columbia Object Image Library</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fcoil20-umap.png" alt="coil20 UMAP (spca)"></td>
<td align="center"><img src="img%2Fexamples%2Fcoil20-tsne.png" alt="coil20 t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="coil100">coil100<a class="anchor" aria-label="anchor" href="#coil100"></a>
</h3>
<p>The <a href="http://www.cs.columbia.edu/CAVE/software/softlib/coil-100.php" class="external-link">COIL-100
Columbia Object Image Library</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fcoil100-umap.png" alt="coil100 UMAP (spca)"></td>
<td align="center"><img src="img%2Fexamples%2Fcoil100-tsne.png" alt="coil100 t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="swiss-roll-isoswiss">swiss roll (isoswiss)<a class="anchor" aria-label="anchor" href="#swiss-roll-isoswiss"></a>
</h3>
<p>The <a href="http://web.mit.edu/cocosci/isomap/datasets.html" class="external-link">Swiss
Roll</a> data used in Isomap. A famous dataset, but perhaps not that
representative of typical real world datasets. t-SNE is known to not
handle this well, but UMAP makes an impressive go at unfolding it.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fisoswiss-umap.png" alt="swiss UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fisoswiss-tsne.png" alt="swiss t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="mnist">mnist<a class="anchor" aria-label="anchor" href="#mnist"></a>
</h3>
<p>The <a href="http://yann.lecun.com/exdb/mnist/" class="external-link">MNIST database of
handwritten digits</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fmnist-umap.png" alt="mnist UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fmnist-tsne.png" alt="mnist t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="fashion">fashion<a class="anchor" aria-label="anchor" href="#fashion"></a>
</h3>
<p>The <a href="https://github.com/zalandoresearch/fashion-mnist" class="external-link">Fashion MNIST
database</a>, images of fashion objects.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Ffashion-umap.png" alt="fashion UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Ffashion-tsne.png" alt="fashion t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="kuzushiji-kmnist">kuzushiji (KMNIST)<a class="anchor" aria-label="anchor" href="#kuzushiji-kmnist"></a>
</h3>
<p>The <a href="https://github.com/rois-codh/kmnist" class="external-link">Kuzushiji MNIST
database</a>, images of cursive Japanese handwriting.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fkuzushiji-umap.png" alt="kuzushiji UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fkuzushiji-tsne.png" alt="kuzushiji t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="norb">norb<a class="anchor" aria-label="anchor" href="#norb"></a>
</h3>
<p>The <a href="https://cs.nyu.edu/~ylclab/data/norb-v1.0-small/" class="external-link">small
NORB dataset</a>, pairs of images of 50 toys photographed at different
angles and under different lighting conditions.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fnorb-umap.png" alt="norb UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fnorb-tsne.png" alt="norb t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="cifar10">cifar10<a class="anchor" aria-label="anchor" href="#cifar10"></a>
</h3>
<p>The <a href="https://www.cs.toronto.edu/~kriz/cifar.html" class="external-link">CIFAR-10
dataset</a>, consisting of 60000 32 x 32 color images evenly divided
across 10 classes (e.g. airplane, cat, truck, bird). t-SNE was applied
to CIFAR-10 in the <a href="http://jmlr.org/papers/v15/vandermaaten14a.html" class="external-link">Barnes-Hut t-SNE
paper</a>, but in the main paper, only the results after passing through
a convolutional neural network were published. t-SNE on the original
pixel data was only given in the <a href="https://lvdmaaten.github.io/publications/misc/Supplement_JMLR_2014.pdf" class="external-link">supplementary
information (PDF)</a> which is oddly hard to find a link to via JMLR or
the article itself, and in the less widely-cited <a href="https://arxiv.org/abs/1301.3342" class="external-link">preliminary investigation into BH
t-SNE</a>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fcifar10-umap.png" alt="cifar10 UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fcifar10-tsne.png" alt="cifar10 t-SNE"></td>
</tr></tbody>
</table>
<p>There is an outlying purple cluster (which isn’t easy to see) in the
center bottom of the UMAP plot. I see the same thing in the Python
implementation, so I don’t think this is a bug in <code>uwot</code>
(although I also said that in a previous version of this page, and it
turned out there <em>was</em> a bug in <code>uwot</code>. The current
result really is closer to the Python version now, though). That cluster
of images is of some automobiles but they all seem to be variations of
the same image. The same cluster is present in the t-SNE plot (bottom
left), but is more comfortably close to the rest of the data.</p>
<p>The existence of these near-duplicates in CIFAR-10 doesn’t seem to
have been widely known or appreciated until quite recently, see for
instance <a href="https://twitter.com/colinraffel/status/1030532862930382848" class="external-link">this
twitter thread</a> and <a href="https://arxiv.org/abs/1806.00451" class="external-link">this
paper by Recht and co-workers</a>. Such manipulations are in line with
the practice of data augmentation that is popular in deep learning, but
you would need to be aware of it to avoid the test set results being
contaminated. These images seem like a good argument for applying UMAP
or t-SNE to your dataset as a way to spot this sort of thing.</p>
<p>As the visualization of CIFAR-10 isn’t very successful with UMAP or
t-SNE, here are some results using the activations of a convnet, similar
to that used in the BH t-SNE paper. For the convnet, I used a <a href="https://keras.io/" class="external-link">keras</a> implementation, taken from the <a href="https://appliedmachinelearning.blog/2018/03/24/achieving-90-accuracy-in-object-recognition-task-on-cifar-10-dataset-with-keras-convolutional-neural-networks/" class="external-link">Machine
Learning in Action blog</a>. Features were Z-scaled as carried out in
the blog, but I used 100 epochs and a batch size of 128 and I also used
the <a href="https://www.cs.toronto.edu/~tijmen/csc321/slides/lecture_slides_lec6.pdf" class="external-link">RMSprop
(PDF)</a> optimizer favored in the <a href="https://www.manning.com/books/deep-learning-with-python" class="external-link">Deep
Learning with Python book</a> (with <code>lr=1e-4</code> and
<code>decay=1e-6</code>). Without any data augmentation, this gave a
test set error of <code>0.1655</code>, slightly lower than the test set
result given in the BH t-SNE paper (which used a different architecture
and without the benefit of an extra 4 years of deep learning research).
After retraining with all 60000 images, the flattened output of the
final max-pool layer was used, giving 2048 features (the BH t-SNE paper
network had 1024 output activations). UMAP and t-SNE results are below.
For the UMAP results, I used the t-UMAP settings with scaled PCA
initialization.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fcifar10act_umap.png" alt="cifar10 activations t-UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fcifar10act_tsne.png" alt="cifar10 activations t-SNE"></td>
</tr></tbody>
</table>
<p>Results don’t look quite as good as those in the BH t-SNE paper, but
they are still an improvement. The orange cluster of automobiles remains
as an outlier, even in the activation space. You can also see it in the
BH t-SNE paper in the lower image in Figure 5 (orange cluster at the
bottom, slightly left of center).</p>
</div>
<div class="section level3">
<h3 id="tasic2018">tasic2018<a class="anchor" aria-label="anchor" href="#tasic2018"></a>
</h3>
<p>The <code>tasic2018</code> dataset is a transcriptomics dataset of
mouse brain cell RNA-seq data from the <a href="http://celltypes.brain-map.org/rnaseq/mouse" class="external-link">Allen Brain Atlas</a>
(originally reported by <a href="http://dx.doi.org/10.1038/s41586-018-0654-5" class="external-link">Tasic and
co-workers</a>. There is gene expression data for 14,249 cells from the
primary visual cortex, and 9,573 cells from the anterior lateral motor
cortex to give a dataset of size n = 23,822 overall. Expression data for
45,768 genes were obtained in the original data, but the dataset used
here follows the pre-processing treatment of <a href="https://doi.org/10.1101/453449" class="external-link">Kobak and Berens</a> which applied
a normalization and log transformation and then only kept the top 3000
most variable genes.</p>
<p>The data can be generated from the Allen Brain Atlas website and
processed in Python by following the instructions in this <a href="https://github.com/berenslab/rna-seq-tsne/blob/master/demo.ipynb" class="external-link">Berens
lab notebook</a>. I output the data to CSV format for reading into R and
assembling into a data frame with the following extra exporting
code:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-log3k.csv"</span>, logCPM, delimiter<span class="op">=</span><span class="st">","</span>)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-areas.csv"</span>, tasic2018.areas, delimiter<span class="op">=</span><span class="st">","</span>, fmt <span class="op">=</span> <span class="st">"</span><span class="sc">%d</span><span class="st">"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-genes.csv"</span>, tasic2018.genes[selectedGenes], delimiter<span class="op">=</span><span class="st">","</span>, fmt<span class="op">=</span><span class="st">'</span><span class="sc">%s</span><span class="st">'</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-clusters.csv"</span>, tasic2018.clusters, delimiter<span class="op">=</span><span class="st">","</span>, fmt<span class="op">=</span><span class="st">'</span><span class="sc">%d</span><span class="st">'</span>)</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-cluster-names.csv"</span>, tasic2018.clusterNames, delimiter<span class="op">=</span><span class="st">","</span>, fmt<span class="op">=</span><span class="st">'</span><span class="sc">%s</span><span class="st">'</span>)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>np.savetxt(<span class="st">"path/to/allen-visp-alm/tasic2018-cluster-colors.csv"</span>, tasic2018.clusterColors, delimiter<span class="op">=</span><span class="st">","</span>, fmt<span class="op">=</span><span class="st">'</span><span class="sc">%s</span><span class="st">'</span>)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Ftasic2018-umap.png" alt="tasic2018 UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Ftasic2018-tsne.png" alt="tasic2018 t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="macosko2015">macosko2015<a class="anchor" aria-label="anchor" href="#macosko2015"></a>
</h3>
<p>Another <a href="https://doi.org/10.1016/j.cell.2015.05.002" class="external-link">transcriptomics data
set</a>, used as an example in <a href="https://github.com/pavlin-policar/openTSNE" class="external-link">openTSNE</a>. This
contains data for 44,808 cells from the mouse retina.</p>
<p>The raw data was fetched similarly to this <a href="https://github.com/hemberg-lab/scRNA.seq.datasets/blob/master/bash/macosko.sh" class="external-link">shell
script from the Hemberg Lab</a> and then the data was prepared using the
<a href="https://github.com/pavlin-policar/openTSNE/blob/master/examples/prepare_macosko_2015.ipynb" class="external-link">openTSNE
notebook by Pavlin Policar</a>. Similarly to the <code>tasic2018</code>
dataset, data was log normalized and the 3,000 most variable genes were
retained.</p>
<p>I exported the data (without the Z-scaling and PCA dimensionality
reduction), as a CSV file, e.g.:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>np.savetxt(<span class="st">"/path/to/macosko2015/macosko2015-log3k.csv"</span>, x, delimiter<span class="op">=</span><span class="st">","</span>)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co"># Use these as column names</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>np.savetxt(<span class="st">"/path/to/macosko2015/macosko2015-genenames.csv"</span>, data.T.columns.values[gene_mask].astype(<span class="bu">str</span>), delimiter<span class="op">=</span><span class="st">","</span>, fmt <span class="op">=</span> <span class="st">"</span><span class="sc">%s</span><span class="st">"</span>)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>np.savetxt(<span class="st">"/path/to/macosko2015/macosko2015-clusterids.csv"</span>, cluster_ids.values.astype(<span class="bu">int</span>), delimiter<span class="op">=</span><span class="st">","</span>, fmt <span class="op">=</span> <span class="st">"</span><span class="sc">%d</span><span class="st">"</span>)</span></code></pre></div>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fmacosko2015-umap.png" alt="macosko2015 UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fmacosko2015-tsne.png" alt="macosko2015 t-SNE"></td>
</tr></tbody>
</table>
<p>Results are not that impressive looking (especially for UMAP). I
include this example because it’s very common in scRNA-seq packages to
apply PCA to the gene expression data before UMAP. In this case, I have
reduced the dimensionality down to 100. After that transformation,
results look very different (with much better separation of
clusters):</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fmacosko2015pca100-umap.png" alt="macosko2015pca100 UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fmacosko2015pca100-tsne.png" alt="macosko2015pca100 t-SNE"></td>
</tr></tbody>
</table>
<p>The good news is that results look better and the nearest neighbor
search step in both t-SNE and UMAP will go faster, which is often why
this is carried out. The bad news is that you now have to justify the
biological basis for the PCA pre-processing, which has clearly changed
the relationship between the different measurements. And why 100
dimensions?</p>
<p>In most datasets I’ve seen PCA pre-processing doesn’t have such a
noticeable effect, which is why I like this dataset.</p>
</div>
<div class="section level3">
<h3 id="s-curve-with-a-hole-scurvehole">S-curve With a Hole (scurvehole)<a class="anchor" aria-label="anchor" href="#s-curve-with-a-hole-scurvehole"></a>
</h3>
<p>Used by the dimensionality reduction method <a href="https://github.com/YingfanWang/PaCMAP" class="external-link">PaCMAP</a>. It is generated
by the script at <a href="https://github.com/YingfanWang/PaCMAP/blob/master/experiments/run_experiments.py" class="external-link uri">https://github.com/YingfanWang/PaCMAP/blob/master/experiments/run_experiments.py</a>.
script mentioned It’s an S-curve with a hole in the middle, and so
should be amenable to being unfolded into 2D. Similar to
<code>isowiss</code>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fscurvehole-umap.png" alt="scurvehole UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fscurvehole-tsne.png" alt="scurvehole t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="mammoth">mammoth<a class="anchor" aria-label="anchor" href="#mammoth"></a>
</h3>
<p>A 3D point cloud of a <a href="https://3d.si.edu/object/3d/mammuthus-primigenius-blumbach:341c96cd-f967-4540-8ed1-d3fc56d31f12" class="external-link">mammoth
at the Smithsonian</a>. A 50,000 point sample was used used in <a href="https://pair-code.github.io/understanding-umap/" class="external-link">Understanding
UMAP</a> and is available for download from the <a href="https://github.com/PAIR-code/understanding-umap" class="external-link">github repo</a>.
There is also a 10,000 point sample available in the same repo, but I
use the 50,000 point version.</p>
<p>As far as I can tell, it was first used for dimensionality reduction
by Max Noichl at <a href="https://github.com/MNoichl/UMAP-examples-mammoth-/blob/9e82eb3ee5b99020d74e99d2060856d49e8b9f85/umammoth.ipynb" class="external-link uri">https://github.com/MNoichl/UMAP-examples-mammoth-/blob/9e82eb3ee5b99020d74e99d2060856d49e8b9f85/umammoth.ipynb</a>.
Following the example in that notebook, the colors in this dataset were
generated by assigning 12 clusters to the 3D data via the <a href="https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html" class="external-link">Agglomerative
Clustering</a> method in <code>sklearn</code>.</p>
<p><code>mammoth</code> is highly associated (at least in my mind) with
the <a href="https://github.com/YingfanWang/PaCMAP" class="external-link">PaCMAP</a> method.
The [PaCMAP paper](<a href="https://jmlr.org/papers/v22/20-1061.html" class="external-link">paper</a> makes
extensive use of it.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fmammoth-umap.png" alt="mammoth UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fmammoth-tsne.png" alt="mammoth t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="hierarchical">hierarchical<a class="anchor" aria-label="anchor" href="#hierarchical"></a>
</h3>
<p>Another dataset used in the <a href="https://jmlr.org/papers/v22/20-1061.html" class="external-link">PaCMAP paper</a>, but
for which I was unable to find the code to generate it in the PaCMAP
repo. The paper describes it thusly: ’consists of 5 “macro clusters”
that each have 5 “meso clusters,” and each of the meso subclusters has 5
“micro clusters.” Thus, there are 5x5x5 = 125 micro clusters total. We
colored each point based on its true macro clusters, shading each meso
cluster.”. Get all of that? There’s more detail in Appendix B of the
paper and I think I was able to turn that into some Python that probably
generated the correct data. I think it’s correct. I will find somewhere
to upload this code eventually.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fhierarchical-umap.png" alt="hierarchical UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fhierarchical-tsne.png" alt="hierarchical t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="spheres">spheres<a class="anchor" aria-label="anchor" href="#spheres"></a>
</h3>
<p>A dataset from the <a href="https://arxiv.org/abs/1906.00722" class="external-link">topological autoencoders</a>
paper. Another synthetic hierarchical dataset, with 10 high-dimensional
spheres embedded in a larger sphere. Most dimensionality reduction
methods that focus on local measures like nearest neighbors have trouble
with these sorts of datasets, where representing the global structure
and scale is challenging.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fspheres-umap.png" alt="spheres UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fspheres-tsne.png" alt="spheres t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="newsgroups-ng20">20 Newsgroups (ng20)<a class="anchor" aria-label="anchor" href="#newsgroups-ng20"></a>
</h3>
<p>The <a href="http://qwone.com/~jason/20Newsgroups/" class="external-link">20 Newsgroups</a>
dataset, containing a sample of Usenet posts (ask your parents or maybe
grandparents at this point) from 20 different newsgroups. For more
details see the <a href="https://jlmelville.github.io/uwot/articles/sparse-data-example.html">Sparse
Data Article</a>.</p>
<p>I refer to this as <code>ng20</code> a lot due to being unable to
begin any variables with a number.</p>
<p>As many dimensionality reduction methods are unable to handle sparse
data, I also sometimes use the dense version of this dataset,
<code>ng20pacmap</code> which can be found in numpy format at the PaCMAP
repo at <a href="https://github.com/YingfanWang/PaCMAP/tree/master/data" class="external-link uri">https://github.com/YingfanWang/PaCMAP/tree/master/data</a>
as <code>20NG.npy</code> (i.e. in numpy format). Labels are also
available there at <code>20NG_labels.npy</code> If I had to guess I
would say that the initial data was subjected to TF-IDF followed by PCA
down to 100 dimensions to convert it from sparse to dense format, but I
don’t know for sure, nor do I know what pre-processing or normalization
was carried out.</p>
<p>The image below is for <code>ng20pacmap</code>.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Fng20pacmap-umap.png" alt="ng20pacmap UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Fng20pacmap-tsne.png" alt="ng20pacmap t-SNE"></td>
</tr></tbody>
</table>
</div>
<div class="section level3">
<h3 id="tomoradar">Tomoradar<a class="anchor" aria-label="anchor" href="#tomoradar"></a>
</h3>
<p>Used in the related papers <a href="https://arxiv.org/abs/2110.02573" class="external-link">T-SNE Is Not Optimized to Reveal
Clusters in Data</a> and <a href="https://arxiv.org/abs/2108.08003" class="external-link">Stochastic Cluster Embedding</a>
and originally from <a href="https://www.mdpi.com/2072-4292/9/1/58" class="external-link">UAV-Borne Profiling Radar
for Forest Research</a>. A link to the data can be found via the <a href="https://github.com/rozyangno/sce" class="external-link">SCE repo</a>. This is a dataset
of microwave radar data from forest canopy. This is a large dataset in
terms of number of items (~120,000) and number of features (~8,000).
This makes it a pain to process, and neither UMAP nor t-SNE do a good
job of separating out the three main classes, but both <code>uwot</code>
and the Python UMAP package take a surprisingly long time to carry out
the spectral initialization, so this could be a good example dataset
when timing and evaluating initialization methods.</p>
<table class="table">
<colgroup>
<col width="51%">
<col width="48%">
</colgroup>
<thead><tr class="header">
<th align="center">UMAP</th>
<th align="center">t-SNE</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center"><img src="img%2Fexamples%2Ftomoradar-umap.png" alt="tomoradar UMAP"></td>
<td align="center"><img src="img%2Fexamples%2Ftomoradar-tsne.png" alt="tomoradar t-SNE"></td>
</tr></tbody>
</table>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Melville.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
